### ISSUE TITLE
SignatureDoesNotMatch for some parts, when `maxConcurrentParts > 1`

### ISSUE THREAD

**Issue by pehrlich** [TTLabs/EvaporateJS#205]
It looks like there are intermittent errors as shown below, which I've been able to aggravate by adding a 2-second `sleep` to my `to_sign` response, and turning the concurrency up to the default (5).

This begins to happen non-deterministically after the first few parts, and can recover later.  In one example, a 17 part file had parts 7-14 fail with the following:

```xml
<Error>
<Code>SignatureDoesNotMatch</Code>
<Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message>
<AWSAccessKeyId>xxxx</AWSAccessKeyId>
<StringToSign>PUT



x-amz-date:Wed, 17 Aug 2016 00:52:15 GMT
/my-bucket/folder1/folder2/file.ext?partNumber=8&uploadId=8FpMEOJNmS63oUt2HNUZeuhQZw3r8OuPFfekrm9toiSjHvyQOMUi2_zLZmi9wZmie2yGc4jHgKxMmXAYV30l6xQ25pzdq7hZZXPyMWWJoSZc9DU4_hlQLidJmysQyjun</StringToSign>
<SignatureProvided>jn6PX+fJbi4pTRgCWh0TujMO2qI=</SignatureProvided>
<StringToSignBytes>50 55 54 0a 0a 0a 0a 78 2d 61 6d 7a 2d 64 61 74 65 3a 57 65 64 2c 20 31 37 20 41 75 67 20 32 30 31 36 20 30 30 3a 35 32 3a 31 35 20 47 4d 54 0a 2f 73 69 6d 75 6c 61 74 6f 72 2d 73 74 61 67 69 6e 67 2f 67 65 6e 65 72 69 63 73 2f 70 72 62 2f 68 61 6e 64 5f 64 69 66 66 75 73 65 2e 74 69 66 3f 70 61 72 74 4e 75 6d 62 65 72 3d 38 26 75 70 6c 6f 61 64 49 64 3d 38 46 70 4d 45 4f 4a 4e 6d 53 36 33 6f 55 74 32 48 4e 55 5a 65 75 68 51 5a 77 33 72 38 4f 75 50 46 66 65 6b 72 6d 39 74 6f 69 53 6a 48 76 79 51 4f 4d 55 69 32 5f 7a 4c 5a 6d 69 39 77 5a 6d 69 65 32 79 47 63 34 6a 48 67 4b 78 4d 6d 58 41 59 56 33 30 6c 36 78 51 32 35 70 7a 64 71 37 68 5a 5a 58 50 79 4d 57 57 4a 6f 53 5a 63 39 44 55 34 5f 68 6c 51 4c 69 64 4a 6d 79 73 51 79 6a 75 6e</StringToSignBytes>
<RequestId>7BD19BEFF51C77C8</RequestId>
<HostId>VzI7VxjtYLMbda9blO3VexfcFtTWe6KPbsMRDgwMnMPchlV0goikGIWgaeNO+RuRvmahCoSe4/Y=</HostId>
</Error>
```

Could there be a race condition in the code which handles the different signatures?  What would be a good way to pursue debugging?

**Comment by bikeath1337** [comment]
Thanks for the report. Not aware of any race conditions here as concurrency is one of the features that's been heavily tested. The best way to debug would be to compare the signature calculation parts from Evaporate with what AWS is using. If you enable logging, you'll see a lot of noise but you could also just add your own logging statements to hone in on the cause.

**Comment by davidferril** [comment]
+1

**Comment by fuzzybro-otoy** [comment]
Same, it's been a major issue for us. Mostly happens on big files (10+ GB).
Except the thing can't recover. Evaporate tries to upload all parts and when it reaches 100% upload just fails. The bigger the value of `maxConcurrentParts` the more it fails.

**Comment by bikeath1337** [comment]
Will look into it tomorrow. Probably introduced in 1.5.0, so 1.47 may be an option for now.

It would be helpful to know which condition triggers it, if anyone has more insight before I take a look.

**Comment by bikeath1337** [comment]
I identified the places where the "in progress" count was not maintained correctly and fixed it on master. Would any of you be able to try it out before publishing?

**Comment by bikeath1337** [comment]
This should have been addressed in 1.5.3.
